Mathlib/Data/Matroid/IndepAxioms.lean

```diff
@@ -0,0 +1,560 @@
+    (maximality : ∀ X, X ⊆ E → ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+    (support : ∀ B, Base B → B ⊆ E) :
+    (Matroid.ofBase E Base exists_base base_exchange maximality support).Base = Base := by
+  ext; simp
+
+/-- A constructor for a matroid using the independence axioms for infinite matroids. -/
+protected def ofIndep (E : Set α) (Indep : Set α → Prop) (h_empty : Indep ∅)
+    (h_subset : ∀ ⦃I J⦄, Indep J → I ⊆ J → Indep I)
+    (h_aug : ∀⦃I B⦄, Indep I → I ∉ maximals (· ⊆ ·) (setOf Indep) →
+      B ∈ maximals (· ⊆ ·) (setOf Indep) → ∃ x ∈ B \ I, Indep (insert x I))
+    (h_maximal : ∀ X, X ⊆ E → ExistsMaximalSubsetProperty Indep X)
+    (h_support : ∀ I, Indep I → I ⊆ E) : Matroid α :=
+  Matroid.ofBase E (· ∈ maximals (· ⊆ ·) (setOf Indep))
+  ( by
+      obtain ⟨B, ⟨hB,-,-⟩, hB₁⟩ := h_maximal E rfl.subset ∅ h_empty (empty_subset _)
+      exact ⟨B, ⟨hB, fun B' hB' hBB' ↦ hB₁ ⟨hB', empty_subset _,h_support B' hB'⟩ hBB'⟩⟩ )
+  ( by
+      rintro B B' ⟨hB, hBmax⟩ ⟨hB',hB'max⟩ e he
+      have hnotmax : B \ {e} ∉ maximals (· ⊆ ·) (setOf Indep)
+      { simp only [mem_maximals_setOf_iff, diff_singleton_subset_iff, not_and, not_forall,
+          exists_prop, exists_and_left]
+        exact fun _ ↦ ⟨B, hB, subset_insert _ _, by simpa using he.1⟩ }
+
+      obtain ⟨f,hf,hfB⟩ := h_aug (h_subset hB (diff_subset B {e})) hnotmax ⟨hB',hB'max⟩
+      simp only [mem_diff, mem_singleton_iff, not_and, not_not] at hf
+
+      have hfB' : f ∉ B := by (intro hfB; obtain rfl := hf.2 hfB; exact he.2 hf.1)
+
+      refine' ⟨f, ⟨hf.1, hfB'⟩, by_contra (fun hnot ↦ _)⟩
+      obtain ⟨x,hxB, hind⟩ :=  h_aug hfB hnot ⟨hB, hBmax⟩
+      simp only [mem_diff, mem_insert_iff, mem_singleton_iff, not_or, not_and, not_not] at hxB
+      obtain rfl := hxB.2.2 hxB.1
+      rw [insert_comm, insert_diff_singleton, insert_eq_of_mem he.1] at hind
+      exact not_mem_subset (hBmax hind (subset_insert _ _)) hfB' (mem_insert _ _) )
+  ( by
+      rintro X hXE I ⟨hB, hB, hIB⟩ hIX
+      obtain ⟨J, ⟨hJ, hIJ, hJX⟩, hJmax⟩ := h_maximal X hXE I (h_subset hB.1 hIB) hIX
+      obtain ⟨BJ, hBJ⟩ := h_maximal E rfl.subset J hJ (h_support J hJ)
+      refine' ⟨J, ⟨⟨BJ,_, hBJ.1.2.1⟩ ,hIJ,hJX⟩, _⟩
+      · exact ⟨hBJ.1.1, fun B' hB' hBJB' ↦ hBJ.2 ⟨hB',hBJ.1.2.1.trans hBJB', h_support _ hB'⟩ hBJB'⟩
+      simp only [maximals, mem_setOf_eq, and_imp, forall_exists_index]
+      rintro A B' (hBi' : Indep _) - hB'' hIA hAX hJA
+      simp only [mem_setOf_eq, and_imp] at hJmax
+      exact hJmax (h_subset hBi' hB'') hIA hAX hJA )
+  ( fun B hB ↦ h_support B hB.1 )
+
+@[simp] theorem ofIndep_apply (E : Set α) (Indep : Set α → Prop) (h_empty : Indep ∅)
+    (h_subset : ∀ ⦃I J⦄, Indep J → I ⊆ J → Indep I)
+    (h_aug : ∀⦃I B⦄, Indep I → I ∉ maximals (· ⊆ ·) (setOf Indep) →
+      B ∈ maximals (· ⊆ ·) (setOf Indep) → ∃ x ∈ B \ I, Indep (insert x I))
+    (h_maximal : ∀ X, X ⊆ E → ExistsMaximalSubsetProperty Indep X)
+    (h_support : ∀ I, Indep I → I ⊆ E) :
```

Comment:
A trick to save having to restate all this is that you can write 
```suggestion
    (h_subset h_aug h_maximal h_support) :
```
That's because `Matroid.ofIndep` determines their types from `E` and `Indep`. (I'd have included `h_empty` to if I had also selected that line before writing this suggestion.)